tools/mage/end-to-end.go

```
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func (EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.Build)
+}
+
+// Prepare prepares the server for running end to end tests.
+func (EndToEnd) Prepare() {
+	mg.Deps(EndToEnd.prepareDB, EndToEnd.prepareBuild)
+}
+
+// DBDump performs a database dump to the .cache folder.
+func (EndToEnd) DBDump() error {
+	if mg.Verbose() {
+		fmt.Printf("Execute database dump\n")
+	}
+	_ = os.MkdirAll(".cache", 0755)
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", testDatabaseName, "--insecure")
+}
+
+// DBRestore restores the database using a previously generated dump.
+func (EndToEnd) DBRestore() error {
+	mg.Deps(Js.Deps, Dev.DBStart)
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
+	}
+	return sh.Run("node", filepath.Join("tools", "mage", "scripts", "restore-db-dump.js"), "--db", testDatabaseName)
+}
+
+// Cypress runs the Cypress end-to-end tests.
+func (e EndToEnd) Cypress() error {
+	mg.Deps(Js.Deps, EndToEnd.WaitUntilReady)
+	if mg.Verbose() {
+		fmt.Println("Running Cypress E2E Tests")
+	}
+	return e.runCypress("run")
+}
+
+// CypressOpen runs the Cypress end-to-end tests in interactive mode.
+func (e EndToEnd) CypressOpen() error {
+	mg.Deps(Js.Deps)
+	if mg.Verbose() {
+		fmt.Println("Running Cypress in interactive mode")
+	}
+	return e.runCypress("open")
+}
+
+// StartTestStack starts TTS in end-to-end test configuration.
+func (EndToEnd) StartTestStack() error {
+	mg.Deps(Js.Build)
+	if mg.Verbose() {
+		fmt.Println("Starting stack in end-to-end test configuration")
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	return runGoMute("./cmd/ttn-lw-stack", "start")
+}
+
+// WaitUntilReady waits until the web endpoints become available. For CI use.
+func (EndToEnd) WaitUntilReady() error {
+	if mg.Verbose() {
+		fmt.Println("Waiting for the stack to be ready...")
+	}
+	for i := 0; i < 100; i++ {
+		resp, _ := http.Get(getTestURL() + "/oauth")
```

Do not discard errors, my earlier comment about this is not addressed.

---

Contrarily, my [reply to your comment](https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442592462) was not addressed by you.

---

This function will block for more 100 seconds if e.g. `getTestURL` returns wrong URL or host's networking is messed up/different from what we expect and no error will ever be produced.
That is not how quality software is made.
Please either just use the solution I proposed or come up with your own, but this should not happen.